### PR TITLE
Add docker compose support for development and production environment

### PR DESCRIPTION
**Pull request by suvash**
I recently thought of jumping in and helping out with this project where possible. I use Docker and Docker Compose a lot, so I thought of starting out with building consistent environments. (as installing GADL is such a pain among others)

I have updated the Readme with instructions on usage. One can test out the staging environment locally as well.

No stress about having to merge/reject this right away. My intention is to start a discussion with a PR that works. 🍻 

Few things to note:
- Usage instructions add in Readme.
- I've included a separate `requirements.txt`, so that it wouldn't affect current requirements, but this can be sorted later.
- Staging can be tested out by running locally with the following changes in `compose.staging.yml`
```
# in caddy > environment
      - HOST_NAME=localhost
      - TLS_EMAIL=self_signed
# in caddy > ports
        - "2015:2015"

```
, and visiting `https://localhost:2015`

### CODE REVIEW

File: compose/web/bin/start-web [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+export PGPASSWORD="$POSTGRES_PASSWORD"
+
+# Setup Simpletable entries
+cat sql/simpletables/*.sql | psql -w -U "$POSTGRES_USERNAME"  -h "$POSTGRES_HOST"
+
+# Migrate
+python ./manage.py migrate --no-input
+
+# Import data for tables
+cat sql/*.sql | psql -w -U "$POSTGRES_USERNAME" -h "$POSTGRES_HOST"
```
Comment (suvash) [inline compose/web/bin/start-web]: @cliftonmcintosh Line 18 - Line25 runs the database loading via sql files as well as runs django migrations. This is run within this `start-web` script before the django server starts. (So, basically whenever this container boots up and runs `start-web` the migrations/db restores are run.)
When a new sql file is added to the folders, one can simply `make stop` and `make start` which will load up all the new files.

Comment (cliftonmcintosh) [reply compose/web/bin/start-web]: @suvash Thanks! I look forward to trying it. I should be able to run through it this weekend. Sorry for the delay, but weekdays are busy for me now.

**Comment by suvash** [comment]
@cliftonmcintosh I can help out with the switch if needed. But, it's important for you to feel a bit comfortable with docker(and docker compose) before we get there. This PR otherwise, has almost everything needed to get things running on production already.

When running locally, you can set the `APP_ENV=dev make ....`. Running `APP_ENV=dev make` without any targets defaults to printing out some basic documentation as well. Like you mentioned, `start` target starts all the necessary bit and gets the web server running. Also, do run `stop` afterwards, otherwise the containers keep running in the background and consume resources.

Maybe, it's a good idea for me to demo the tooling over a video chat. (something like https://appear.in)

File: compose/web/bin/start-web (hunk 1 above)
Comment (suvash) [reply compose/web/bin/start-web]: No worries. Let me know if something doesn't work etc. afterwards. 🍻

**Comment by cliftonmcintosh** [comment]
@suvash 

I have tried the set up using `APP_ENV=dev make start`.

Dependencies appear to be retrieved, but the startup script stops when trying to run the entrypoint file:

```
ERROR: for web  Cannot start service web: oci runtime error: container_linux.go:247: starting container process caused "exec: \"/usr/src/app/compose/web/entrypoint.sh\": permission denied"
```

**Comment by suvash** [comment]
@cliftonmcintosh That's a bit puzzling. If I understand this properly, the file is not permitted to execute (but, I'm pretty sure that the `+x` bit is set properly).
- Can I ask what OS you're running (most likely recent versions of docker/compose)
- Can you run docker without sudo ? (`docker run --rm hello-world` is a good test)
- Do you get back results(some ids) when running `id -u` & `id -g` on a shell ?
- If possible can you share the logs you get when running `make tail-logs APP_ENV=dev` ?

**Comment by cliftonmcintosh** [comment]
@suvash 

It was operator error. When I applied a [patch based on the PR](https://patch-diff.githubusercontent.com/raw/Code4Nepal/nepalmap_app/pull/191.patch),  `entrypoint.sh` did not get set as executable. That may simply be because of the way I applied the patch. I've solved that by making it executable. The containers start up and everything works fine.

**Comment by cliftonmcintosh** [comment]
@suvash 

I'm happy to approve and merge this pull request as a way to allow for easier development. 

As far as staging/test or production environments go, the effort to change how we deploy is **currently** not a high priority for me because:

1. My time is limited right now.
2. My focus (and interest) is in making more data available.
3. We don't deploy very often.
4. The current deployments take a few minutes at most. (I run two commands on the server, and I'm done.)
5. Given the simplicity of the current deployment, the effort to switch how we do deployments outweighs the benefit over the short term.

However, if you have an interest in taking on deployments, we can talk with @ravinepal to see about getting another pair of servers spun up to let you demonstrate or set up deployments using docker.

File: .gitignore [hunk 2]
```diff
@@ -4,5 +4,4 @@ node_modules/
 .idea/
 debug
 .vagrant
-/static/
```
Comment (cliftonmcintosh) [inline .gitignore]: IMO, having `static` here is a better choice than putting a `.gitignore` file inside the static directory.

File: static/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
[the remark this answers is not in the document]
Comment (suvash) [reply static/.gitignore]: 👍

File: .gitignore (hunk 2 above)
Comment (suvash) [reply .gitignore]: 👍

**Comment by suvash** [comment]
@cliftonmcintosh ah, good to know that it works. Pushed a commit for using `.keep` instead. Having just one `.gitignore` is understandable.

For now, I'm fine with starting out with development environment. If we feel like using docker based setup for deployments in the future(let's say sometime next year), maybe after you've had time to look into docker and friends, I can jump right in.

**Comment by cliftonmcintosh** [comment]
@suvash 

Thank you for this work! I'm excited about this lowering the barrier for anyone who might be interested in contributing.

**Comment by suvash** [comment]
@cliftonmcintosh 🍻

**Pull request merged**